### ISSUE TITLE
investigate scheduler approaches

### ISSUE THREAD

**Issue by bengreenier** [Azure/platform-chaos#17]
We desire a way to schedule chaotic events via a hosted service, to remove the dependency on manual intervention as currently exists with https://github.com/Azure/platform-chaos-cli.

This tracks the work we need to do to identify if there is a possible existing solution out there that may be coupled with the work we've already done to provide these features.

Current possibilities include:

+ Azure Scheduler
+ Azure Logic apps
+ Lightweight azure function (timer trigger)

**Comment by bengreenier** [comment]
depends on #16 - as we'll definitely want to call out to that api from whatever tool or solution we define for scheduling

**Comment by mindlessroman** [comment]
Starting thoughts

## Azure Scheduler
* "run jobs [...] on any schedule, making it ideal for recurring actions like cleaning up logs, kicking off backups, and other maintenance tasks"
* Price: $13.99 per unit / per month
* Unlimited number of executions
* Max run frequency is once a minute
* Lots of customization how a job is run and good UI - once v recurring
* Scheduler has history/logging
* [Getting Started docs](https://docs.microsoft.com/en-us/azure/scheduler/scheduler-get-started-portal)
* [API Docs](https://docs.microsoft.com/en-us/rest/api/scheduler/)

## Logic Apps
* Out of the box connections to a lot of kinds of services outside Azure
* Pricing per Action/Connection - ($0.000025, $0.000125 respectively)
* [Getting Started docs](https://docs.microsoft.com/en-us/azure/logic-apps/quickstart-create-first-logic-app-workflow)
* [Example Tutorial with scheduled task](https://docs.microsoft.com/en-us/azure/logic-apps/tutorial-build-schedule-recurring-logic-app-workflow) - decision trees

## Azure Functions
* Available free tier - up to 1million executions
* Execution time: 400,000 GB-s
* Lots of language support
* [General Documentation](https://docs.microsoft.com/en-us/azure/azure-functions/)
* [Timer tutorial as an Azure Function](https://docs.microsoft.com/en-us/azure/azure-functions/functions-create-scheduled-function)

## Azure Batch
* "You don’t need to write your own work queue, dispatcher, or monitor. Batch gives you this as a service."
* [Pricing](https://azure.microsoft.com/en-us/pricing/details/batch/)
* [General Docs](https://docs.microsoft.com/en-us/azure/batch/)

## Jenkins
* Not Azure-exclusive, but is [supported](https://docs.microsoft.com/en-us/azure/jenkins/)
* Maybe more configuration-heavy and not as CLI friendly under the hood?
* [Docs](https://jenkins.io/doc/)

**Comment by bengreenier** [comment]
thanks for including those @mindlessroman! I'd be interested to hear what you recommend we use after doing some of that research.

**Comment by mindlessroman** [comment]
Discussed this a lot more in person during standup - the current front runner is Azure Scheduler. But some more questions/thoughts were raised:

How does each platform handle providing a level of randomness?
What has a good/the best UI for a partner/user to configure?
Scheduler - what does the max frequency entail?

**Comment by mindlessroman** [comment]
# Scheduler
## Randomness:
Trying to instantiate some chaos within a scheduled window would likely require some more coding gymnastics - doesn't seem to be built in.  Probably not impossible 😅 More thinking would need to be done, but one thought: using an azure function (or something similar) to generate a time in specified window, then inputting that into the scheduler. 

## UI
Creating a quick scheduler was fairly simple and there are some user-configurable fields.
![2018-09-04 16_59_42-action settings - microsoft azure](https://user-images.githubusercontent.com/7254163/45057885-e4fffd00-b064-11e8-8cbe-a76215114402.png)
![2018-09-04 17_00_28-action settings - microsoft azure](https://user-images.githubusercontent.com/7254163/45057956-2a242f00-b065-11e8-93d0-f25065d6827e.png)
![2018-09-04 17_01_00-authentication - microsoft azure](https://user-images.githubusercontent.com/7254163/45057976-3f995900-b065-11e8-8c94-4b85c50cdca0.png)
![2018-09-04 17_01_25-retry policy - microsoft azure](https://user-images.githubusercontent.com/7254163/45057998-5344bf80-b065-11e8-94e8-341a8c955ab0.png)
![2018-09-04 17_01_46-error action settings - microsoft azure](https://user-images.githubusercontent.com/7254163/45058007-58a20a00-b065-11e8-8467-56ca2854bddf.png)

UI for timing (one-time v recurring) is pretty self-explanatory.
![2018-09-04 17_02_33-schedule - microsoft azure](https://user-images.githubusercontent.com/7254163/45058051-78393280-b065-11e8-9574-12a6d634cca9.png)
![2018-09-04 17_02_55-schedule - microsoft azure](https://user-images.githubusercontent.com/7254163/45058067-80916d80-b065-11e8-9e55-0224dbb5db83.png)

## History 
Doesn't mention _who_ triggered it, but does manage history by scheduled task (in this example `hktestingscheduler`). 
![image](https://user-images.githubusercontent.com/7254163/45058497-e2060c00-b066-11e8-98f6-09a16c86d7ae.png)
![image](https://user-images.githubusercontent.com/7254163/45058545-01049e00-b067-11e8-8854-4f5fb290bcc9.png)
 
## Frequency
The [docs](https://docs.microsoft.com/en-us/azure/scheduler/scheduler-limits-defaults-errors) mention the limit on frequency. There will be a 409 error code if one tries to manually make a frequency that's higher than once a minute. 

## Conclusion
May be a better choice for chaos that occurs at predictable, settable times. ie chaotic actions with 'set' timing, not chaotically timed so-to-speak. But may be able to hack a solution for more time-chaos.

**Comment by mindlessroman** [comment]
More details for the other options to come...

**Comment by mindlessroman** [comment]
At this point, there doesn't seem to be one of these that offers a single solution to the problem we're hoping to address. In that case, we'll probably need have a blended solution that includes a few of these. Logic Apps definitely wins with the UI - very configurable, layman-readable and is the leading candidate in the Azure-generated solutions for something that is timed chaotically. Scheduler would be a good candidate for the predictably timed chaos, but with slightly less layman-readable UI. The UI's still very configurable, but it lacks the same visual engagement/flow that Logic Apps has. 

Azure Functions would likely need to be used in our solution for generating a "random" time in a specified window.

**Comment by mindlessroman** [comment]
We had another chat last week. Using Logic Apps may be over engineering the problem - the leading solution is to invoke the chaos via CLI and take care of the randomly generated time that way. Scheduler itself doesn't care how a job is scheduled, so we should randomize at the layer ahead of Scheduler.
